### DISCUSSION TITLE
Import functions from other ps1 files referencing it

### DISCUSSION THREAD

**Discussion by mdurini-swisscom** [PowerShell/PowerShell#24120] [General]
### Summary of the new feature / enhancement

Hello, let me clarify the title of this question. I write some ps1 scripts in VS Code that i need to provide to other users: I would like to find a way to keep all the functions in a separated file, reference it in the final script to be able to test it in VS Code and, once I'm ready to send this file to the user, copy all the functions that I used in the script from the file where they are defined. I know that I could write a Powershell Module, but this will imply and I will need to distribute it somehow and it will not be easy. Thank you in advance

**Comment by kilasuit** [comment]
This is not really a feature request that I think the team would look at as we already as part of the PowerShell Gallery and the Script capabilities there can distribute a script that has dependencies and have that script auto download dependencies as needed.

I'm going to convert this to a discussion for others to provide some input on how you could achieve this

**Comment by rhubarb-geek-nz** [comment]
I am inferring that you want to do the equivalent of classic C application static linking where the single a.out contains all the _used_ library functions.

So lets say you have a library directory with functions in separate files.....

```
lib\foo.ps1
lib\bar.ps1
lib\baz.ps1
lib\frobincate.ps1
```

Those can be used independently in other scripts, eg

```
. lib\foo.ps1
. lib\bar.ps1

foo this
bar that
```

So that lets you develop without code duplication

Then to publish your final script you generate the public version replacing the dot-sourced lines with the contents of the original file from the lib folder

```
function foo
{ things that make you go foo }

function bar
{ goto the bar }

foo this
bar that
```

**Reply by mdurini-swisscom** [reply]
Hello @rhubarb-geek-nz yes you're right, now I would like to find a way to automate the last step, something like a VSCode extension or a build/compile script that will ingest the script with the references and generate a new script with all the functions that I used

**Comment by kasini3000** [comment]
import-module "${PSScriptRoot}\lib\foo.ps1"
func_name_in_foo -param1 'value1'

**Reply by mdurini-swisscom** [reply]
the idea is to be able to "build" the file that i want to distribute with all the functions that I used, copying them from the ps1 files

**Comment by IISResetMe** [comment]
You can use PowerShell's own parser to [parse both scripts](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.language.parser.parsefile?view=powershellsdk-7.4.0) and then [extract the relevant function definitions](https://vexx32.github.io/2018/12/20/Searching-PowerShell-Abstract-Syntax-Tree/) from the shared script:

userscript.ps1:
```pwsh
param(
  [string]$SomeInputParameter
)

# dot-source shared functions scripts during development
. C:\hardcoded\path\to\shared.ps1

$stuff = Get-Stuff -Param $SomeInputParameter
Write-Host "Done getting stuff, now we'll process it"
$stuff |Process-Stuff
```

shared.ps1:
```pwsh
function Get-Stuff {
  param($Param)
  
  <# ... #>
}

function Process-Stuff {
  param([Parameter(ValueFromPipeline)]$InputObject)
  
  <# ... #>
}

function Unrelated-Function {
  param()
}
```

packager.ps1:
```pwsh
param(
  [Parameter(Mandatory)]
  [ValidateScript({ $_ |Test-Path -PathType Leaf })]
  [string]$EndUserScriptPath,

  [ValidateScript({ $_ |Test-Path -PathType Leaf })]
  [string]$SharedScriptPath = $('C:\hardcoded\path\to\shared.ps1')
)

# parse the script
$parserErrors = @()
$scriptAST = [System.Management.Automation.Language.Parser]::ParseFile((Convert-Path $EndUserScriptPath), [ref]$null, [ref]$parserErrors)

if ($parserErrors) {
  throw "Errors occurred while parsing script!"
}

# locate all command invocations in the script
$commandInvocations = $scriptAST.FindAll({
  return $args[0] -is [System.Management.Automation.Language.CommandAst]
}, $true)
$commandNames = $commandInvocations |ForEach-Object {
  $_.CommandElements[0].Value
}

# parse the shared functions script and locate matching function definitions
$sharedAST = [System.Management.Automation.Language.Parser]::ParseFile((Convert-Path $SharedScriptPath), [ref]$null, [ref]$null)

$matchingFunctionDefinitions = $sharedAst.FindAll({
  return $args[0] -is [System.Management.Automation.Language.FunctionDefinitionAst]
}, $false)

# write new script consisting of the scripts param block + body, 
# but with the dot-source operation removed and the function definitions inserted in its place
$dotSourceStatement = $scriptAST.Find({
  return $args[0] -is [System.Management.Automation.Language.CommandAst] -and 
             $args[0].InvocationOperator -eq 'Dot' -and 
            ($args[0].CommandElements[0].Value |Split-Path -Leaf) -eq ($SharedScriptPath|Split-Path -Leaf)
}, $true)

if (-not $dotSourceStatement) {
  throw "failed to locate dot-source statement"
}

# figure out where to replace the dot-source statement with the function definitions, then do the remove/insert and return the resulting string
$insertAt = $dotSourceStatement.Extent.StartOffset
$cutLength = $dotSourceStatement.Extent.EndOffset - $insertAt

return $scriptAST.Extent.Text.Remove($insertAt, $cutLength).Insert($insertAt, $(
  @($matchingFunctionDefinitions |ForEach-Object { $_.Extent.Text }) -join ([Environment]::NewLine * 2)
))
```

Then invoke it like this:

```
PS ~> .\packager.ps1 path\to\userscript.ps1 -SharedScriptPath path\to\shared.ps1
param(
  [string]$SomeInputParameter
)

# dot-source shared functions scripts during development
function Get-Stuff {
  param($Param)

  <# ... #>
}

function Process-Stuff {
  param([Parameter(ValueFromPipeline)]$InputObject)

  <# ... #>
}

function Unrelated-Function {
  param()
}

$stuff = Get-Stuff -Param $SomeInputParameter
Write-Host "Done getting stuff, now we'll process it"
$stuff |Process-Stuff
```

Since `packager.ps1` just outputs a string, you can write the results straight to a new file:
```
.\packager.ps1 path\to\userscript.ps1 -SharedScriptPath path\to\shared.ps1 |Set-Content path\to\userscript_standalone.ps1
```

You probably wanna add some additional error handling, but this should get you started at least.

**Comment by hrxn** [comment]
> [..]  I know that I could write a Powershell Module, but this will imply and I will need to distribute it somehow and it will not be easy.  [..]

How so?
Why would this imply that you have to distribute anything? I don't think so.

A module can be a single `.psm1` file, use the `Export-ModuleMember` function at the end of that file to define what will be exported, everything not explicitly mentioned there will be local only. You don't need a manifest file or anything.

Because if I understand that correctly

> [..] once I'm ready to send this file to the user, copy all the functions that I used in the script from the file where they are defined. 
[..]

You end up with a single file for distribution, containing all function definitions etc. anyway?

**Reply by rhubarb-geek-nz** [reply]
> You end up with a single file for distribution, containing all function definitions etc. anyway?

That is not the goal. The goal is to deliver ready to run scripts with all the contained functions, not modules that have to be imported in order to be used or put in a special modules directory.
